### ISSUE TITLE
Nullability of reference types in type of parameter 'b' of 'lambda expression' doesn't match the target delegate 'Action<KeysetPaginationBuilder<Category?>>' (possibly because of nullability attributes).  csharp(CS8622)

### ISSUE THREAD

**Issue by ugumerie** [mrahhal/MR.AspNetCore.Pagination#6]
**Version Used:** 2.0.0

I'm making using of Roslynator extension in VSCode for code refactoring, fixing, analyzing etc. 
**Nullable** is also enabled
**Target framework:** 6.0

**Code snippet**
```
....
return await _paginationService.KeysetPaginateAsync(
      dbContext.Categories!,
      b => b.Descending(x => x!.CreatedOn),
      async id => await _categories.FindAsync(Guid.Parse(id)),
      query => query.ProjectToType<GetKeysetPagedCategoriesResponse>()
 );
```
**Compilation warnings:**
`Nullability of reference types in type of parameter 'b' of 'lambda expression' doesn't match the target delegate 'Action<KeysetPaginationBuilder<Category?>>' (possibly because of nullability attributes).  csharp(CS8622)`

`The type 'SoftPurse.Modules.ServiceCatalog.Core.Catalog.Aggregates.Category?' cannot be used as type parameter 'T' in the generic type or method 'IPaginationService.KeysetPaginateAsync<T, TOut>(IQueryable<T>, Action<KeysetPaginationBuilder<T>>, Func<string, Task<T>>, Func<IQueryable<T>, IQueryable<TOut>>, KeysetQueryModel)'. Nullability of type argument 'SoftPurse.Modules.ServiceCatalog.Core.Catalog.Aggregates.Category?' doesn't match 'class' constraint. csharp(CS8634)`

**Note:** **Category** mention here is the entity.
It is being inferred as `await _paginationService.KeysetPaginateAsync<Category?, GetKeysetPagedCategoriesResponse>(...)`

**Comment by mrahhal** [comment]
I'm not familiar with Roslynator so I can't comment much on that, but the usage of null suppressions (`!`) is a red flag. Nullable properties are not supported in keysets in general, as mentioned in the [caveats in the KeysetPagination package](https://github.com/mrahhal/MR.EntityFrameworkCore.KeysetPagination/blob/main/docs/caveats.md#null). So that might be the cause of that analyzer warning.

**Comment by ugumerie** [comment]
Thank you very much. I just worked on explicitly passed the `T` and `TOut`.
Here is the code like so:
```
....
 return await _paginationService.KeysetPaginateAsync<Category, GetKeysetPagedCategoriesResponse>(
        _categories.Where(c => c.IsActive),
        b => b.Descending(x => x.CreatedOn),
        async id => await _categories.FindAsync(Guid.Parse(id)),
        query => query.ProjectToType<GetKeysetPagedCategoriesResponse>(),
        queryModel: new KeysetQueryModel
        {
            First = true,
            Size = 10
        }
    );
```
**Warning**: `Possible null reference return` on ` async id => await _categories.FindAsync(Guid.Parse(id))`
I still get a `Possible null reference return` warning on the `Func<string, Task<T>> getReferenceAsync` delegate. since `FindAsync()` returns an object or null. But the code works perfectly and does the pagination. Just that the warning is not good for the eyes :)

**PaginationService code snippet:**
....
else if (queryModel.After != null)
{
	var reference = await getReferenceAsync(queryModel.After);
	keysetContext = query.KeysetPaginate(builderAction, KeysetPaginationDirection.Forward, reference);
}
else if (queryModel.Before != null)
{
	var reference = await getReferenceAsync(queryModel.Before);
	// This is a special case.
	// If reference is null (maybe the entity was deleted from the database), we want to
	// always return the first page, so enforce a Forward direction.
	var direction = reference != null ? KeysetPaginationDirection.Backward : KeysetPaginationDirection.Forward;
	keysetContext = query.KeysetPaginate(builderAction, direction, reference);
}
....

**Question on PaginationService code snippet:**
Is there a way you could also check whether the reference returned from ` queryModel.After` is null, just like you did for `queryModel.Before` such that the `getReferenceAsync` delegate parameter would look like this - `Func<string, Task<T?>> getReferenceAsync` where T could be null?

**Comment by ugumerie** [comment]
Aha, great, thank you. Will do just that.

**Comment by mrahhal** [comment]
So regarding this:
https://github.com/mrahhal/MR.AspNetCore.Pagination/blob/ad81c6220b248514031d87d755a0d7db6076c2eb/src/MR.AspNetCore.Pagination/PaginationService.cs#L268-L281

The reason the `After` block didn't have the same logic is because it's already Forward in direction, so the behavior does actually match the `Before` block on nulls. `Before` is a special case because we want to enforce Forward. In any case, this issue still stands to resolve the nullability mismatch.

**Comment by mrahhal** [comment]
Should be fixed in v2.0.1.
